proposed/simple-csharp-programs.md

```diff
@@ -0,0 +1,461 @@
+
+This leads to some questions:
+
+* Are multiple file supported? That is, can I define a `server.cs` file that is my web service, and a `data.cs` that handles some data manipulation that is called from code inside of `server.cs`?
+* If multiple files are supported, what are the guardrails? How can people know if they made a mistake and fix it?
+
+It's also quite likely that some developers will have several programs all as single C# files in the same directory. That means that `dotnet run` would have to understand this and give an appropriate error message about specifying a specific C# file. However, this means inspecting the files to see if they are being used via `#load` directive (or something similar).
+
+Consider the folowing two subfolders.
+
+Directory One:
+
+```
+reboot-server.cs
+check-server-status.cs
+```
+
+Directory Two:
+
+```
+reboot-server.cs
+utils.cs
+```
+
+In Directory Two, the `utils.cs` file is used by `reboot-server.cs`. A developer may expect to be able to just run `dotnet run` in that directory and have the program execute. But if they are in Directory One, they would have to be explicit about which file they are running.
+
+### Coding against a framework alternative syntax
+
+The walkthrough proposes the `#!` directive `#!/path/to/dotnet-aspnet`, but the .NET CLI doesn't have a command to "pull in the ASP.NET Core references", so it wouldn't map to something like that.
+
+An alternative could be to extend `#r` like so:
+
+* `#r "framework:aspnet"`
+* `#r "sdk:aspnet"`
+* etc.
+
+This could cleanly separate the "tell the compiler which references to pull in" world from the "short-cut having to use `dotnet`" world like so:
+
+```csharp
+#!/path/to/dotnet
+#r "framework:aspnet"
+
+var app = WebApplication.CreateBuilder(args);
+app.MapGet("/", () => "Hello World");
+await app.RunAsync();
+```
+
+### `#!` and its semantics
+
+The `#!` directives in the walkthrough all point at `/path/to/dotnet`. This path may be different depending on your development OS (e.g., Fedora).
+
+What this means is that it identifies the .NET environment and implies a default set of references available in the file. However, it does not bring in other things like ASP.NET or other frameworkreferences. It is a contextual, "this is how I run" sort of directive.
+
+It could be extended/changed in a few ways:
+
+* `/path/to/dotnet run` makes it explicit that the command is to run the file
+* `/path/to/dotnet-aspnet` is an alternative to `#!aspnet` that brings ASP.NET Core into scope for the file (substitute a different framework/sdk here)
+
+In effect, the `#!` directive's semantics could span a spectrum of "I need the path and command to literally execute this file" towards "I add context that affects the references available to this file".
+
+We'll need to decide if they can affect the set of references available in the file or not.
+
+### Using `#!` to point to `dotnet` or point to `env`
```

Comment:
The `#!` will make it possible to execute these files on Linux just using the filename.

Is it desired/possible on Windows too? Maybe by having a dedicated extension?

Reply:
Maybe, although `#!` really isn't a convention on Windows.